File: backend/routes/configurations.js
const express = require('express');
const { v4: uuidv4 } = require('uuid');
const crypto = require('crypto-js');
const router = express.Router();

// POST /api/configurations/sessions - Create new configuration session
router.post('/sessions', async (req, res) => {
  try {
    const { supabase } = req;
    const { userEmail, userName, userPhone, userType = 'customer' } = req.body;

    // Generate session token
    const sessionToken = crypto.lib.WordArray.random(32).toString();

    const { data: session, error } = await supabase
      .from('configuration_sessions')
      .insert({
        session_token: sessionToken,
        user_email: userEmail,
        user_name: userName,
        user_phone: userPhone,
        user_type: userType,
        current_step: 'chassis_selection',
        session_status: 'active',
        expires_at: new Date(Date.now() + 30 * 24 * 60 * 60 * 1000) // 30 days
      })
      .select()
      .single();

    if (error) {
      console.error('Session creation error:', error);
      return res.status(500).json({ error: 'Failed to create session' });
    }

    res.status(201).json({
      success: true,
      data: {
        sessionId: session.id,
        sessionToken: session.session_token,
        currentStep: session.current_step,
        expiresAt: session.expires_at
      }
    });

  } catch (error) {
    console.error('Create session route error:', error);
    res.status(500).json({ error: 'Internal server error' });
  }
});

// GET /api/configurations/sessions/:sessionId - Get session state
router.get('/sessions/:sessionId', async (req, res) => {
  try {
    const { supabase } = req;
    const { sessionId } = req.params;

    const { data: session, error } = await supabase
      .from('configuration_sessions')
      .select(`
        *,
        configuration_selections (
          *,
          vehicle_options (
            option_code,
            option_name
          )
        )
      `)
      .eq('id', sessionId)
      .single();

    if (error) {
      console.error('Session fetch error:', error);
      return res.status(404).json({ error: 'Session not found' });
    }

    // Check if session is expired
    if (new Date(session.expires_at) < new Date()) {
      return res.status(410).json({ error: 'Session expired' });
    }

    res.json({
      success: true,
      data: session
    });

  } catch (error) {
    console.error('Get session route error:', error);
    res.status(500).json({ error: 'Internal server error' });
  }
});

// PUT /api/configurations/sessions/:sessionId - Update session
router.put('/sessions/:sessionId', async (req, res) => {
  try {
    const { supabase } = req;
    const { sessionId } = req.params;
    const { 
      currentStep, 
      selectedChassisId, 
      selectedBodyId, 
      selectedFuelType,
      basePrice,
      optionsPrice,
      totalPrice 
    } = req.body;

    const updateData = {};
    if (currentStep) updateData.current_step = currentStep;
    if (selectedChassisId) updateData.selected_chassis_id = selectedChassisId;
    if (selectedBodyId) updateData.selected_body_id = selectedBodyId;
    if (selectedFuelType) updateData.selected_fuel_type = selectedFuelType;
    if (basePrice !== undefined) updateData.base_price = basePrice;
    if (optionsPrice !== undefined) updateData.options_price = optionsPrice;
    if (totalPrice !== undefined) updateData.total_price = totalPrice;

    const { data: session, error } = await supabase
      .from('configuration_sessions')
      .update(updateData)
      .eq('id', sessionId)
      .select()
      .single();

    if (error) {
      console.error('Session update error:', error);
      return res.status(500).json({ error: 'Failed to update session' });
    }

    res.json({
      success: true,
      data: session
    });

  } catch (error) {
    console.error('Update session route error:', error);
    res.status(500).json({ error: 'Internal server error' });
  }
});

// POST /api/configurations/sessions/:sessionId/selections - Add selection to session
router.post('/sessions/:sessionId/selections', async (req, res) => {
  try {
    const { supabase } = req;
    const { sessionId } = req.params;
    const { selectionType, selectedItemId, selectedItemCode, quantity = 1, unitPrice, totalPrice } = req.body;

    const { data: selection, error } = await supabase
      .from('configuration_selections')
      .insert({
        session_id: sessionId,
        selection_type: selectionType,
        selected_item_id: selectedItemId,
        selected_item_code: selectedItemCode,
        quantity,
        unit_price: unitPrice,
        total_price: totalPrice,
        is_valid: true
      })
      .select()
      .single();

    if (error) {
      console.error('Selection creation error:', error);
      return res.status(500).json({ error: 'Failed to add selection' });
    }

    res.status(201).json({
      success: true,
      data: selection
    });

  } catch (error) {
    console.error('Add selection route error:', error);
    res.status(500).json({ error: 'Internal server error' });
  }
});

// DELETE /api/configurations/sessions/:sessionId/selections/:selectionId - Remove selection
router.delete('/sessions/:sessionId/selections/:selectionId', async (req, res) => {
  try {
    const { supabase } = req;
    const { sessionId, selectionId } = req.params;

    const { error } = await supabase
      .from('configuration_selections')
      .delete()
      .eq('id', selectionId)
      .eq('session_id', sessionId);

    if (error) {
      console.error('Selection deletion error:', error);
      return res.status(500).json({ error: 'Failed to remove selection' });
    }

    res.json({
      success: true,
      message: 'Selection removed successfully'
    });

  } catch (error) {
    console.error('Remove selection route error:', error);
    res.status(500).json({ error: 'Internal server error' });
  }
});

// POST /api/configurations/sessions/:sessionId/validate - Validate current configuration
router.post('/sessions/:sessionId/validate', async (req, res) => {
  try {
    const { supabase } = req;
    const { sessionId } = req.params;

    // Get current session
    const { data: session, error: sessionError } = await supabase
      .from('configuration_sessions')
      .select('*')
      .eq('id', sessionId)
      .single();

    if (sessionError) {
      return res.status(404).json({ error: 'Session not found' });
    }

    const validationResults = [];

    // Validate chassis-body compatibility
    if (session.selected_chassis_id && session.selected_body_id) {
      const { data: compatibility, error: compatError } = await supabase
        .from('chassis_body_compatibility')
        .select('*')
        .eq('base_vehicle_id', session.selected_chassis_id)
        .eq('body_config_id', session.selected_body_id)
        .eq('is_compatible', true)
        .single();

      if (compatError || !compatibility) {
        validationResults.push({
          type: 'compatibility',
          status: 'error',
          message: 'Selected chassis and body are not compatible',
          code: 'INCOMPATIBLE_CHASSIS_BODY'
        });
      } else {
        validationResults.push({
          type: 'compatibility',
          status: 'passed',
          message: 'Chassis and body are compatible'
        });
      }
    }

    // Validate fuel type compatibility
    if (session.selected_chassis_id && session.selected_fuel_type) {
      const { data: fuelCompat, error: fuelError } = await supabase
        .from('chassis_fuel_compatibility')
        .select('*')
        .eq('base_vehicle_id', session.selected_chassis_id)
        .eq('fuel_type_options.fuel_code', session.selected_fuel_type)
        .eq('availability_status', 'Available');

      if (fuelError || !fuelCompat || fuelCompat.length === 0) {
        validationResults.push({
          type: 'fuel_compatibility',
          status: 'error',
          message: 'Selected fuel type is not available for this chassis',
          code: 'INCOMPATIBLE_FUEL_TYPE'
        });
      } else {
        validationResults.push({
          type: 'fuel_compatibility',
          status: 'passed',
          message: 'Fuel type is compatible with chassis'
        });
      }
    }

    // Log validation results
    for (const result of validationResults) {
      await supabase
        .from('configuration_validations')
        .insert({
          session_id: sessionId,
          validation_type: result.type,
          validation_status: result.status,
          error_code: result.code,
          error_message: result.message
        });
    }

    const hasErrors = validationResults.some(r => r.status === 'error');

    res.json({
      success: true,
      data: {
        isValid: !hasErrors,
        validationResults
      }
    });

  } catch (error) {
    console.error('Validation route error:', error);
    res.status(500).json({ error: 'Internal server error' });
  }
});

// POST /api/configurations/save - Save configuration
router.post('/save', async (req, res) => {
  try {
    const { supabase } = req;
    const { sessionId, configurationName, userEmail, isFavorite = false } = req.body;

    // Get session data
    const { data: session, error: sessionError } = await supabase
      .from('configuration_sessions')
      .select(`
        *,
        configuration_selections (*)
      `)
      .eq('id', sessionId)
      .single();

    if (sessionError) {
      return res.status(404).json({ error: 'Session not found' });
    }

    // Generate share token
    const shareToken = crypto.lib.WordArray.random(32).toString();

    // Save configuration
    const { data: savedConfig, error } = await supabase
      .from('saved_configurations')
      .insert({
        configuration_name: configurationName,
        user_email: userEmail,
        session_id: sessionId,
        configuration_data: {
          chassis_id: session.selected_chassis_id,
          body_id: session.selected_body_id,
          fuel_type: session.selected_fuel_type,
          selections: session.configuration_selections
        },
        total_price: session.total_price,
        is_favorite: isFavorite,
        share_token: shareToken
      })
      .select()
      .single();

    if (error) {
      console.error('Save configuration error:', error);
      return res.status(500).json({ error: 'Failed to save configuration' });
    }

    res.status(201).json({
      success: true,
      data: {
        configurationId: savedConfig.id,
        shareToken: savedConfig.share_token,
        shareUrl: `${req.protocol}://${req.get('host')}/shared/${savedConfig.share_token}`
      }
    });

  } catch (error) {
    console.error('Save configuration route error:', error);
    res.status(500).json({ error: 'Internal server error' });
  }
});

// GET /api/configurations/shared/:shareToken - Get shared configuration
router.get('/shared/:shareToken', async (req, res) => {
  try {
    const { supabase } = req;
    const { shareToken } = req.params;

    const { data: config, error } = await supabase
      .from('saved_configurations')
      .select('*')
      .eq('share_token', shareToken)
      .single();

    if (error) {
      console.error('Shared config error:', error);
      return res.status(404).json({ error: 'Configuration not found' });
    }

    res.json({
      success: true,
      data: config
    });

  } catch (error) {
    console.error('Shared configuration route error:', error);
    res.status(500).json({ error: 'Internal server error' });
  }
});

module.exports = router;

File: deployment/src/routes/configurations.py
from flask import Blueprint, jsonify, request
from supabase import create_client, Client
import secrets
from datetime import datetime, timedelta

configurations_bp = Blueprint('configurations', __name__)

# Supabase configuration
SUPABASE_URL = "https://rfctmbpdthtovqkogbol.supabase.co"
SUPABASE_KEY = "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJzdXBhYmFzZSIsInJlZiI6InJmY3RtYnBkdGh0b3Zxa29nYm9sIiwicm9sZSI6ImFub24iLCJpYXQiOjE3NTUwMTc0OTUsImV4cCI6MjA3MDU5MzQ5NX0.RzvfalZtvT1ytxzCsaQaXN53vpgnY9iZ8Z4DiZIWQKU"

supabase: Client = create_client(SUPABASE_URL, SUPABASE_KEY)

@configurations_bp.route('/configurations/sessions', methods=['POST'])
def create_session():
    try:
        data = request.get_json() or {}
        user_type = data.get('userType', 'customer')
        
        # Generate session token
        session_token = secrets.token_hex(32)
        
        # Create session
        session_data = {
            'session_token': session_token,
            'user_type': user_type,
            'current_step': 'chassis_selection',
            'session_status': 'active',
            'expires_at': (datetime.now() + timedelta(days=30)).isoformat()
        }
        
        response = supabase.table('configuration_sessions').insert(session_data).execute()
        
        if response.data:
            session = response.data[0]
            return jsonify({
                'sessionId': session['id'],
                'sessionToken': session['session_token'],
                'userType': session['user_type'],
                'currentStep': session['current_step'],
                'status': session['session_status'],
                'expiresAt': session['expires_at']
            }), 201
        else:
            return jsonify({'error': 'Failed to create session'}), 500
            
    except Exception as e:
        print(f"Session creation error: {e}")
        return jsonify({'error': 'Failed to create configuration session'}), 500

@configurations_bp.route('/configurations/sessions/<session_id>/selections', methods=['POST'])
def create_selection(session_id):
    try:
        data = request.get_json()
        
        # Validate required fields
        required_fields = ['selectionType', 'selectedItemId']
        for field in required_fields:
            if field not in data:
                return jsonify({'error': f'Missing required field: {field}'}), 400
        
        selection_data = {
            'session_id': session_id,
            'selection_type': data['selectionType'],
            'selected_item_id': data['selectedItemId'],
            'selected_item_code': data.get('selectedItemCode'),
            'quantity': data.get('quantity', 1),
            'unit_price': data.get('unitPrice'),
            'total_price': data.get('totalPrice'),
            'is_valid': True
        }
        
        response = supabase.table('configuration_selections').insert(selection_data).execute()
        
        if response.data:
            selection = response.data[0]
            return jsonify({
                'id': selection['id'],
                'sessionId': selection['session_id'],
                'selectionType': selection['selection_type'],
                'selectedItemId': selection['selected_item_id'],
                'selectedItemCode': selection['selected_item_code'],
                'quantity': selection['quantity'],
                'unitPrice': selection['unit_price'],
                'totalPrice': selection['total_price'],
                'isValid': selection['is_valid']
            }), 201
        else:
            return jsonify({'error': 'Failed to create selection'}), 500
            
    except Exception as e:
        print(f"Selection creation error: {e}")
        return jsonify({'error': 'Failed to create selection'}), 500

@configurations_bp.route('/configurations/sessions/<session_id>', methods=['GET'])
def get_session(session_id):
    try:
        response = supabase.table('configuration_sessions').select('*').eq('id', session_id).execute()
        
        if not response.data:
            return jsonify({'error': 'Session not found'}), 404
            
        session = response.data[0]
        return jsonify({
            'sessionId': session['id'],
            'sessionToken': session['session_token'],
            'userType': session['user_type'],
            'currentStep': session['current_step'],
            'status': session['session_status'],
            'selectedChassisId': session.get('selected_chassis_id'),
            'selectedBodyId': session.get('selected_body_id'),
            'basePrice': session.get('base_price'),
            'optionsPrice': session.get('options_price'),
            'totalPrice': session.get('total_price'),
            'expiresAt': session['expires_at']
        })
        
    except Exception as e:
        print(f"Session query error: {e}")
        return jsonify({'error': 'Failed to fetch session'}), 500

